src/Microsoft.IdentityModel.Abstractions/ITelemetryClient.cs

```diff
@@ -0,0 +1,87 @@
+﻿//------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.IdentityModel.Abstractions
+{
+    /// <summary>
+    /// Interface for Telemetry tracking.
+    /// </summary>
+    internal interface ITelemetryClient
+    {
+        /// <summary>
+        /// Gets or sets the application or client ID that telemetry is being sent for.
+        /// </summary>
+        Guid ClientId { get; set; }
```

Comment:
Is it possible that we might want to add more properties like ClientName/ClientDisplayName or EventStatus,...etc?
interface will be hard to change (vs abstract class that can have more methods and not breaking the contract) if this becomes public.

Reply:
Agree with Danny that an interface will be hard to change once it becomes public. You could also consider having a struct/another class (TelemetryContext) to store all the properties of ITelemetryClient. That class would be easy to extend and the interface would remain unchanged.

Reply:
was only internal in the first iteration, now I'll be making it public since we actually want public constructors to pass this type.

IMO interfaces are cleaner and allow folks to better reuse their existing code. I agree they have different challenges with extensibility. IMO I partly mitigated this with EventDetails and it can be further mitigated by interface inheritance. I don't feel super strongly though. I'd rather get it in as an abstract class than spin wheels here. I've went interface->abstract class->interface already. Seems like the team is split. We don't seem to have a team wide guideline so I went with what I ultimately think is the cleaner design. Open to having my mind changed or just changing back to an abstract class to keep moving 😂

Reply:
If the expectation is for users to extend the TelemetryClient then maybe abstract class is a good choice. Interface should be enough if this should cover common functionality. Starting from C# 8, interface seems to be looking more like abstract class "Beginning with C# 8.0, an interface may define a default implementation for members. It may also define static members in order to provide a single implementation for common functionality."